functable.c

```diff
@@ -157,190 +160,195 @@ static void init_functable(void) {
-        ft.longest_match = &longest_match_rvv;
-        ft.longest_match_slow = &longest_match_slow_rvv;
-        ft.slide_hash = &slide_hash_rvv;
+        ft->adler32 = &adler32_rvv;
+        ft->adler32_fold_copy = &adler32_fold_copy_rvv;
+        ft->chunkmemset_safe = &chunkmemset_safe_rvv;
+        ft->chunksize = &chunksize_rvv;
+        ft->compare256 = &compare256_rvv;
+        ft->inflate_fast = &inflate_fast_rvv;
+        ft->longest_match = &longest_match_rvv;
+        ft->longest_match_slow = &longest_match_slow_rvv;
+        ft->slide_hash = &slide_hash_rvv;
     }
 #endif
 
 
     // S390
 #ifdef S390_CRC32_VX
     if (cf.s390.has_vx)
-        ft.crc32 = crc32_s390_vx;
+        ft->crc32 = crc32_s390_vx;
+#endif
+}
+
+#ifdef FUNCTABLE_ATOMIC_STORE
+#  if defined(__GNUC__) || defined(__clang__)
+#    define FUNCTABLE_ASSIGN(VAR, FUNC_NAME) \
+    __atomic_store (&(functable.FUNC_NAME), &(VAR->FUNC_NAME), __ATOMIC_SEQ_CST)
+#  elif defined(_MSC_VER)
+#    define FUNCTABLE_ASSIGN(VAR, FUNC_NAME) \
+    InterlockedExchangePointer((volatile PVOID *)&(functable.FUNC_NAME), (PVOID)(VAR->FUNC_NAME))
+#  endif
+#endif
+#ifndef FUNCTABLE_ASSIGN
+#  define FUNCTABLE_ASSIGN(VAR, FUNC_NAME) \
+    functable.FUNC_NAME = VAR->FUNC_NAME
 #endif
 
+static void apply_functable(struct functable_s *ft) {
     // Assign function pointers individually for atomic operation
-    functable.adler32 = ft.adler32;
-    functable.adler32_fold_copy = ft.adler32_fold_copy;
-    functable.chunkmemset_safe = ft.chunkmemset_safe;
-    functable.chunksize = ft.chunksize;
-    functable.compare256 = ft.compare256;
-    functable.crc32 = ft.crc32;
-    functable.crc32_fold = ft.crc32_fold;
-    functable.crc32_fold_copy = ft.crc32_fold_copy;
-    functable.crc32_fold_final = ft.crc32_fold_final;
-    functable.crc32_fold_reset = ft.crc32_fold_reset;
-    functable.inflate_fast = ft.inflate_fast;
-    functable.insert_string = ft.insert_string;
-    functable.longest_match = ft.longest_match;
-    functable.longest_match_slow = ft.longest_match_slow;
-    functable.quick_insert_string = ft.quick_insert_string;
-    functable.slide_hash = ft.slide_hash;
-    functable.update_hash = ft.update_hash;
+    FUNCTABLE_ASSIGN(ft, adler32);
+    FUNCTABLE_ASSIGN(ft, adler32_fold_copy);
+    FUNCTABLE_ASSIGN(ft, chunkmemset_safe);
+    FUNCTABLE_ASSIGN(ft, chunksize);
+    FUNCTABLE_ASSIGN(ft, compare256);
+    FUNCTABLE_ASSIGN(ft, crc32);
+    FUNCTABLE_ASSIGN(ft, crc32_fold);
+    FUNCTABLE_ASSIGN(ft, crc32_fold_copy);
+    FUNCTABLE_ASSIGN(ft, crc32_fold_final);
+    FUNCTABLE_ASSIGN(ft, crc32_fold_reset);
+    FUNCTABLE_ASSIGN(ft, inflate_fast);
+    FUNCTABLE_ASSIGN(ft, insert_string);
+    FUNCTABLE_ASSIGN(ft, longest_match);
+    FUNCTABLE_ASSIGN(ft, longest_match_slow);
+    FUNCTABLE_ASSIGN(ft, quick_insert_string);
+    FUNCTABLE_ASSIGN(ft, slide_hash);
+    FUNCTABLE_ASSIGN(ft, update_hash);
 }
 
+#define STUB_BODY(RET, FUNC_NAME, ...) \
+    struct functable_s ft; \
```

Comment:
Should this be `static`? Am I missing something, but wouldn't this evaluate the entire `functable` for each function?

Reply:
No.
A stub may be called more than once only if the first call of versioned functions is parallel. To avoid race we need local variable.